### DISCUSSION TITLE
audio & video playback in lags safari

### DISCUSSION THREAD

**Discussion by jubejubess** [jspsych/jsPsych#809] [Q&A]
Hi!

I began testing a smaller experiment in a few different browsers and I found that audio and video playback was very laggy using safari in comparison to other browsers. My experiment plays audio clips (occasionally plays a few videos in between) and participants has to respond with which sound they heard.

I found that sometimes the audio clips had a slight delay or played two clips that overlapped at once - sometimes it was the same audio clip that played twice and overlapped or it was the current audio clip overlapped with next audio clip that was supposed to play after).
The same occurs with video clips as well - sometimes it would cut off or even skip the video clip instead of playing it. 

I tried to change the parameters of the audio clips to play and stay on the screen for a longer amount of time so that the audio could play fully before the participant can answer. This temporarily worked for fixing the audio playback for Safari (audio playback began working the same as it originally did for the other browsers).

However, for the other browsers, this change in the parameters created a delay between the audio & response screens that’s becoming too long (since the other browsers had no issues with audio play in the first place).

I was wondering if there was an alternative way of fixing the media playback for safari that wouldn’t affect the other browsers or if this is just an issue with how Safari handles audio/video stimuli. We don’t mind excluding safari as an browser option if it comes to that, but we’d like to see if we could fix this issue some other way! 

Thank you!

**Comment by jkhartshorne** [comment]
@jubejubess could you say a bit more about what device you were using? In [this recent study](https://github.com/jspsych/jsPsych/discussions/806), audio lag in Safari was pretty good for jsPsych. Also, you may be interested in #806.

**Comment by jodeleeuw** [comment]
Hi @jubejubess. It would also be helpful to know what plugin you were using for the playback and how your timeline was set up. It could be that the media files were not preloaded for some reason.

**Comment by jubejubess** [comment]
Hello! Thanks for your responses.

This was tested on both apple desktops (imac) and laptops (macbook pro). The plug-ins used were 'html-keyboard-response' and 'audio-keyboard-response'.

This is how the practice trials in the timeline are set up. The test trials are set up the same, but just that there are more audio files loaded: 

```js
// list of practice stimuli
    var practice_stimuliDL = [
      {stimulus: "audio/bd01.wav", data: {level: '1', dresponse: 'DL', test: 'practice'}},
      {stimulus: "audio/bd09.wav", data: {level: '9', dresponse: 'DL', test: 'practice'}},
  ];

var practice_audioDL = {
                  type: "audio-keyboard-response",
                  stimulus: jsPsych.timelineVariable('stimulus'),
                  choices: jsPsych.NO_KEYS,
                  trial_ends_after_audio: true
                }

            // after audio finishes completely, participant can make a response & data is logged
            var practice_responseDL = {
                  type: "html-keyboard-response",
                  stimulus: "<div style='font-size:25px';'><div style='width: 500px;'><div style='float:left';'><b>ADA</b></div>" +
                  "<div style='font-size:25px';'><div style='width: 500px;'><div style='float:right';'><b>ABA</b></div>",
                  choices: ['leftarrow', 'rightarrow'],
                  response_ends_trial: true,
                  trial_duration: null,
                  data: jsPsych.timelineVariable('data'), 
                  on_finish: function(data){ 
                    if (data.key_press == 37) {
                      data.resp = 0;
                    } else if (data.key_press == 39) {
                      data.resp = 1;
                    }
                  }
                }

            // practice trial procedure */
            var practice_procedureDL = {
                  timeline: [practice_audioDL , practice_responseDL],
                  timeline_variables: practice_stimuliDL,
                  randomize_order: true
                }

// same process for audio stimuli... 
// [half] list of audio stimuli
var audio_stimuliDL = [
      { stimulus: "audio/bd01.wav", data: {level: '1', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd01.wav", data: {level: '1', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd01.wav", data: {level: '1', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd01.wav", data: {level: '1', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd01.wav", data: {level: '1', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd02.wav", data: {level: '2', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd02.wav", data: {level: '2', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd02.wav", data: {level: '2', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd02.wav", data: {level: '2', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd02.wav", data: {level: '2', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd02.wav", data: {level: '2', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd02.wav", data: {level: '2', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd02.wav", data: {level: '2', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd03.wav", data: {level: '3', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd03.wav", data: {level: '3', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd03.wav", data: {level: '3', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd03.wav", data: {level: '3', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd03.wav", data: {level: '3', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd03.wav", data: {level: '3', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd03.wav", data: {level: '3', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd03.wav", data: {level: '3', dresponse: 'DL', test: 'test'}},
      { stimulus: "audio/bd03.wav", data: {level: '3', dresponse: 'DL', test: 'test'}},

if(randomint > 239){
          timeline.push(practicetrialinstructionsDL, practice_procedureDL, testclick, audio_procedureDL);
        } else if(randomint < 239){
          timeline.push(practicetrialinstructionsDR, practice_procedureDR, testclick, audio_procedureDR);
        }


// in the jspsych init
preload_audio: ["audio/bd01.wav", "audio/bd02.wav", "audio/bd03.wav", "audio/bd04.wav", "audio/bd05.wav", "audio/bd06.wav", "audio/bd07.wav", "audio/bd08.wav", "audio/bd09.wav" ], //preload audio
        use_webaudio: false,
```

**Reply by jodeleeuw** [reply]
I'm not sure what the culprit is here. My best guess is that it might be because the audio files are in `wav` format. Would converting to MP3 be an option? Depending on how big these files are you might be hitting issues with how much data the browser is willing to preload for your page.

**Reply by jkhartshorne** [reply]
@jubejubess @jodeleeuw I'm also wondering about using WebAudio API. In the past, that was more reliable than HTML5, so maybe you want to set it to true?

**Reply by jubejubess** [reply]
I tried converting and testing it out with .mp3 files and setting HTML5 to true, but the delay still occurs unfortunately. The files are quite small about ~29 kb each and the experiment presents about 90 files in total. 

We may just exclude safari as a browser option if it's all too fussy!
